pkg/controller/utils/elasticsearch.go

```diff
@@ -113,69 +112,55 @@ func GetElasticsearchClusterConfig(ctx context.Context, cli client.Client) (*rel
 	return relasticsearch.NewClusterConfigFromConfigMap(configMap)
 }
 
+type ElasticsearchClientCreator func(client client.Client, ctx context.Context, elasticHTTPSEndpoint string) (ElasticClient, error)
+
 type ElasticClient interface {
-	SetILMPolicies(client.Client, context.Context, *operatorv1.LogStorage, string) error
+	SetILMPolicies(context.Context, *operatorv1.LogStorage) error
 }
 
 type esClient struct {
 	client *elastic.Client
-	lock   sync.Mutex
-}
-
-func NewElasticClient() ElasticClient {
-	return &esClient{}
 }
 
-// SetILMPolicies creates ILM policies for each timeseries based index using the retention period and storage size in LogStorage
-func (es *esClient) SetILMPolicies(client client.Client, ctx context.Context, ls *operatorv1.LogStorage, elasticHTTPSEndpoint string) error {
-	es.lock.Lock()
-	if err := es.createElasticClient(client, ctx, elasticHTTPSEndpoint); err != nil {
-		es.lock.Unlock()
-		return err
+func NewElasticClient(client client.Client, ctx context.Context, elasticHTTPSEndpoint string) (ElasticClient, error) {
+	user, password, root, err := getClientCredentials(client, ctx)
+	if err != nil {
+		return nil, err
+	}
+	h := &http.Client{
+		Transport: &http.Transport{TLSClientConfig: &tls.Config{RootCAs: root}},
 	}
-	es.lock.Unlock()
-	policyList := es.listILMPolicies(ls)
-	return es.createOrUpdatePolicies(ctx, policyList)
-}
 
-func (es *esClient) createElasticClient(client client.Client, ctx context.Context, elasticHTTPSEndpoint string) error {
-	if es.client == nil {
-		user, password, root, err := getClientCredentials(client, ctx)
-		if err != nil {
-			return err
-		}
-		h := &http.Client{
-			Transport: &http.Transport{TLSClientConfig: &tls.Config{RootCAs: root}},
-		}
+	options := []elastic.ClientOptionFunc{
+		elastic.SetURL(elasticHTTPSEndpoint),
+		elastic.SetHttpClient(h),
+		elastic.SetErrorLog(logrWrappedESLogger{}),
+		elastic.SetSniff(false),
+		elastic.SetHealthcheck(false),
+		elastic.SetBasicAuth(user, password),
+	}
+	retryInterval, err := time.ParseDuration(ElasticConnRetryInterval)
+	if err != nil {
+		return nil, err
+	}
 
-		options := []elastic.ClientOptionFunc{
-			elastic.SetURL(elasticHTTPSEndpoint),
-			elastic.SetHttpClient(h),
-			elastic.SetErrorLog(logrWrappedESLogger{}),
-			elastic.SetSniff(false),
-			elastic.SetHealthcheck(false),
-			elastic.SetBasicAuth(user, password),
-		}
-		retryInterval, err := time.ParseDuration(ElasticConnRetryInterval)
-		if err != nil {
-			return err
+	var esCli *elastic.Client
+	for i := 0; i < ElasticConnRetries; i++ {
+		esCli, err = elastic.NewClient(options...)
+		if err == nil {
+			break
 		}
+		log.Error(err, "Elastic connect failed, retrying")
+		time.Sleep(retryInterval)
+	}
 
-		var eserr error
-		var esClient *elastic.Client
-		for i := 0; i < ElasticConnRetries; i++ {
+	return &esClient{client: esCli}, nil
```

Comment:
I think this case needs to be handled - say if the `esCli` is never created after `ElasticConnRetries`, we will have to return error.

Reply:
oh right woops thanks for catching that

Reply:
fixed